### PR TITLE
Don't use magic comments, let there be single indented magics

### LINKED ISSUE #248: Get rid of "# NBQAMAGIC" magic
There's got to be a better way than to use this hack...

### LINKED ISSUE #249: indented magics don't work if they're meant to be indented
e.g.

```python
if True:
    %time True
pass
```

This will get transformed to
```python
if True:
# commented out line
pass
```
which isn't valid syntax.

I'm tempted to say that the fix here would be to replace magics with
```
pass  # nbqamagic
```
and then, in the future, have a less hacky solution

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #249 
closes #248 

Previously,
```python
if True:
    %time my_func()
```
would not have worked, because
```python
if True:
    # NBQAMAGIC
```
isn't valid syntax. So instead, now, I'm replace the magic with a hash, and store which original line the hash corresponds to. So here would would get, e.g.
```python
if True:
    pass  # nbqa1509fb
```
, which _is_ valid syntax.

Some of the tests needed updating, because `black` and `flake8` treat `pass` differently than comments. However, IMO their current treatment is better.

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
This still doesn't work for ignoring entire cells though...those probably do need to be commented out entirely.

**Comment by MarcoGorelli** [comment]
OK, current solution is to comment-out cells which are to be ignored, and to just replace inline-magics with `pass`.

Like this, it correctly formats all the notebooks from PyMC3 without errors

**Comment by MarcoGorelli** [comment]
Have merged the conflict - if it's OK with you, I'd be tempted to merge this as it is, make a 0.2.0 release (after which we can start following a bit more of a process, with a changelog and release notes) and then in 0.2.x releases we optimise this / use `print` when we add support for `pylint`

**Review by girip11 — APPROVED** [review]

File: nbqa/save_source.py [hunk 1]
```diff
@@ -39,19 +43,67 @@ def _replace_magics(source: List[str], ignore_cells: List[str]) -> Iterator[str]
...
+        ) or ignore_cell:
+            token = secrets.token_hex(3)
+            if ignore_cell:
+                # Comment out entire cell.
+                replacement = IGNORE_CELL_REPLACEMENT
+                leading_space = ""
+            else:
+                # Replace line with `pass`.
+                replacement = IGNORE_LINE_REPLACEMENT
+                leading_space = " " * (len(line) - len(line.lstrip(" ")))
```
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply nbqa/save_source.py:58]: Nice! I don't know about `itertools.takewhile`, thanks for the suggestion!

**Pull request merged**